src/htm/utils/Random.hpp

```diff
@@ -165,23 +169,21 @@ class Random : public Serializable  {
   friend UInt32 GetRandomSeed();
 private:
   UInt64 seed_;
-  UInt64 steps_ = 0;  //step counter, used in serialization. It is important that steps_ is in sync with number of 
+  mutable UInt64 steps_ = 0;  //step counter, used in serialization. It is important that steps_ is in sync with number of 
   // calls to RNG
-  std::mt19937 gen; //Standard mersenne_twister_engine 64bit seeded with seed_
+  mutable std::mt19937 gen; //Standard mersenne_twister_engine 64bit seeded with seed_
```

Comment:
all `Random` methods can be made const (thread safe), with the exception in mutable here. Result, will work OK, but for effective parallel algorithm, we likely won't have deterministic builds. So when finished, I imagine we run parallel by default, but when stuff breaks, we switch to single thread and deterministic operation.

Reply:
@marty1885 thank you for your advice. What do you think of the mutable in Random here? Shouldn't make too much damage if some collision messed the state. 

but it results in the non-deterministic builds. (Currently we support set seed and deterministic operation for the whole HTM, which is cool for some usages).

Reply:
I'll say that using a shared RNG across threads is not a good idea. Besides making things nondeterministic. It also leads to false sharing and a tremendous amount of cache invalidation. (Even though I do share a RNG across threads in Etaler)

One solution is to have a master RNG, and then you use the master RNG to seed light weight RNGs local to each thread. This will make your results deterministic as long as you have the same # of thread.

Reply:
Update: you can have a master RNG, then use the master RNG to seed a secondary RNG local to each function. Then use the secondary to seed thread-local RNGs. This way you can have deterministic results no matter the amount of threads.

Reply:
In PRs that prepare for this functionality, like #552 , I'm removing dependency on local variables / state. That is a good thing overall. With the suggested approach as above, that is "passing all that is needed as parameters" I need to balance convenience and usability of the public API (for single threaded use), I don't want users who want to call `SDR result = SP.compute(input)` to need to pass 10s parameters (Random, ...). 

So I'll need to figure some compromise. Maybe function overloads for user,parallel? `compute(simple, params)` x `compute(many,needed,for,parallel)` 

Also whole SP.compute is quite bad idea to parallelize, rather some inner methods: inhibition, overlaps,...